src/sonic-host-services/scripts/hostcfgd

```diff
@@ -57,15 +71,210 @@ def obfuscate(data):
+            raise ValueError('Invalid state rendered for feature {}: {}'.format(self.name, target_state))
+        return target_state
+
+
+class FeatureHandler(object):
+    """ Handles FEATURE table updates. """
+
+    SYSTEMD_SYSTEM_DIR = '/etc/systemd/system/'
+    SYSTEMD_SERVICE_CONF_DIR = os.path.join(SYSTEMD_SYSTEM_DIR, '{}.service.d/')
+
+    def __init__(self, config_db, device_config):
+        self._config_db = config_db
+        self._device_config = device_config
+        self._cached_config = {}
+        self.is_multi_npu = device_info.is_multi_npu()
+
+    def handle(self, feature_name, feature_cfg):
+        if not feature_cfg:
+            self._cached_config.pop(feature_name)
+            syslog.syslog(syslog.LOG_INFO, "Deregistering feature {}".format(feature_name))
+            return
+
+        feature = Feature(feature_name, feature_cfg, self._device_config)
+        self._cached_config.setdefault(feature_name, Feature(feature_name, {}))
+
+        # Change auto-restart configuration first.
+        # If service reached failed state before this configuration applies (e.g. on boot)
+        # the next called self.update_feature_state will start it again. If it will fail
+        # again the auto restart will kick-in. Another order may leave it in failed state
+        # and not auto restart.
+        if self._cached_config[feature_name].auto_restart != feature.auto_restart:
+            self.update_feature_auto_restart(feature)
+            self._cached_config[feature_name].auto_restart = feature.auto_restart
+
+        # Enable/disable the container service if the feature state was changed from its previous state.
+        if self._cached_config[feature_name].state != feature.state:
+            if self.update_feature_state(feature):
+                self._cached_config[feature_name].state = feature.state
+            else:
+                self.resync_feature_state(self._cached_config[feature_name])
+
+    def update_all_features_config(self):
+        feature_table = self._config_db.get_table('FEATURE')
+        for feature_name in feature_table.keys():
+            if not feature_name:
+                syslog.syslog(syslog.LOG_WARNING, "Feature is None")
+                continue
+
+            feature = Feature(feature_name, feature_table[feature_name], self._device_config)
+            self._cached_config.setdefault(feature_name, feature)
+
+            self.update_feature_auto_restart(feature)
+
+            self.update_feature_state(feature)
+            self.resync_feature_state(feature)
+
+    def update_feature_state(self, feature):
+        cached_feature = self._cached_config[feature.name]
+        enable = False
+        disable = False
+
+        # Allowed transitions:
+        #  None           -> always_enabled
+        #                 -> always_disabled
+        #                 -> enabled
+        #                 -> disabled
+        #  always_enabled -> always_disabled
```

Comment:
Did you mean?
`always_disabled -> always_disabled`
`always_enabled -> always_enabled`

Reply:
This is something I was wondering and related to comments from yozhao101.

Is a transition always_enabled -> always_disabled - allowed? In the current hostcfgd from master it is.
I also noticed few issues with current hostcfgd with regards to these transitions that I am not sure if they are expected or not, didn't found much info about these new "always*" states.
Transition to the same state is noop. If you could point me to some documents/PR describing the desired behaviour it would help me a lot to understand this.

Reply:
@tahmed-dev

Reply:
Thanks @stepanblyschak! The always_enabled was introduced by @abdosi via [PR:6000](https://github.com/Azure/sonic-buildimage/pull/6000) It is meant to be immutable. It got enhance by @yxieca vi [PR:6067](https://github.com/Azure/sonic-buildimage/pull/6067) to allow the feature to be disabled. Currently `always_disabled` is used to disable feature on deployments that don't support this feature.

Reply:
I see, so I think it works as expected with my changes:
```
admin@r-bulldog-04:~$ show feature status | grep teamd
teamd           always_enabled  enabled
admin@r-bulldog-04:~$ sudo config feature state teamd disabled
Feature 'teamd' state is always enabled and can not be modified
admin@r-bulldog-04:~$ show feature status | grep teamd
teamd           always_enabled  enabled
admin@r-bulldog-04:~$ redis-cli -n 4 hset "FEATURE|teamd" state disabled
(integer) 0
admin@r-bulldog-04:~$ show feature status | grep teamd
teamd           always_enabled  enabled
admin@r-bulldog-04:~$ redis-cli -n 4 hset "FEATURE|teamd" state always_disabled
(integer) 0
admin@r-bulldog-04:~$ show feature status | grep teamd
teamd           always_disabled  enabled
admin@r-bulldog-04:~$ redis-cli -n 4 hset "FEATURE|teamd" state always_enabled
(integer) 0
admin@r-bulldog-04:~$ show feature status | grep teamd
teamd           always_disabled  enabled
admin@r-bulldog-04:~$ redis-cli -n 4 hset "FEATURE|teamd" state enabled
(integer) 0
admin@r-bulldog-04:~$ show feature status | grep teamd
teamd           always_disabled  enabled

```